runtimeconfig/manager.go

```diff
@@ -118,13 +130,108 @@ func (om *Manager) CloseListenerChannel(listener <-chan interface{}) {
 	}
 }
 
-func (om *Manager) loop(ctx context.Context) error {
+func (om *Manager) detectChanges(ctx context.Context) error {
 	if om.cfg.LoadPath == "" {
 		level.Info(om.logger).Log("msg", "runtime config disabled: file not specified")
 		<-ctx.Done()
 		return nil
 	}
 
+	if om.cfg.ChangeDetector == "" || om.cfg.ChangeDetector == Poll {
+		om.pollLoop(ctx, func(error) bool { return false })
+	} else if om.cfg.ChangeDetector == Notify {
+		err := om.notifyLoop(ctx)
+		if err != nil {
+			level.Error(om.logger).Log("msg", "failed to initialize runtime configuration notify loop", "err", err)
+			return err
+		}
+	} else {
+		return errors.Errorf("Unsupported runtime configuration change detector: %s", om.cfg.ChangeDetector)
+	}
+
+	return nil
+}
+
+// notifyLoop utilizes file notifications to reduce the amount of reads it performs
+// it will conditionally resort to poll reading when encountering errors loading the configuration
+func (om *Manager) notifyLoop(ctx context.Context) error {
+	// This is tricky to get right, see the inotify man page and https://github.com/fsnotify/fsnotify/issues/372 for a list of issues
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	// If the file is a symlink follow it, otherwise modifications could be missed
+	fp, err := filepath.EvalSymlinks(om.cfg.LoadPath)
+	if err != nil {
+		return err
+	}
+
+	// Watch the directory rather than the file directly to handle the file's remove/add, since inodes are watched not names
+	dir, _ := filepath.Split(fp)
```

Comment:
The description of `filepath.Split()` says "If there is no Separator in path, Split returns an empty dir", wouldn't it be possible that a user specifies only a file in the same directory, without any `/`?

Reply:
That's a good point. I assumed a full path would be provided, but there's actually no guarantee of that. I'll look into this, thanks.

Reply:
Changed to:
```go
	fp, err := filepath.Abs(symlinkFp)
	if err != nil {
		return err
	}
	// Watch the directory rather than the file directly to handle the file's remove/add, since inodes are watched not names
	dir, _ := filepath.Split(fp)
	if dir == "" {
		return errors.New("Could not determine path to watch for notifications from configured load path")
	}
```